### ISSUE TITLE
Can the P4 Runtime implementation re-order operations within a Write batch?

### ISSUE THREAD

**Issue by antoninbas** [p4lang/PI#300]
We had a discussion on this a while back and I think we agreed on these 2 statements:
1. when the Write RPC returns, all operations in the Write batch have been committed to HW
2. the P4 Runtime implementation is free to re-order operations within a batch when possible

I'm pretty sure I have 1. right, but my memory is a little fuzzy about 2.

**Comment by antoninbas** [comment]
@aghaffarkhah @wmohsin any comments?

**Comment by jafingerhut** [comment]
Some clarifying questions for my understanding:

1. Can a batch consist of arbitrary modification operations on any stateful elements in a P4 data plane?  Or is it perhaps restricted to only be for one table per batch?  (allowing a batch to be over an arbitrary set of P4 data plane stateful elements is obviously more general, and thus probably more useful)

2. What does "when a Write RPC returns" mean from the perspective of a controller?  That it has received a _response_ message from the P4Runtime API server/agent?  If it does mean a response message, that seems to imply that perhaps the controller must wait for this response message before it can issue another write batch that it wants to ensure do not start until after the previous batch is committed to the dataplane, which seems to require perhaps-unnecessary controller-device RTTs in some sequences of operations.

If it instead means "in the local implementation of write batches inside of a P4Runtime API server/agent, it is free to start on a subsequent write batch as soon as it knows that all operations in the previous batch have been committed to the data plane", that enables higher performance control plane updates in some cases (and perhaps a fair number of practical use cases that people would like to be fast).

**Comment by jafingerhut** [comment]
FYI, as far as this statement of yours: "2. the P4 Runtime implementation is free to re-order operations within a batch when possible".  I believe that if you _don't_ allow this, you may be significantly limiting the performance that could otherwise be achieved by a P4 Runtime server/agent implementation, at least for some implementations.

**Comment by antoninbas** [comment]
1. Arbitrary modifications on any subset of P4 objects
2. It means the client indeed has to wait. With the current RPC design, a given client can have at most one Write batch in flight, i.e. I don't believe there is a way to send 2 concurrent batches and guarantee their ordering. I believe we discussed the possibility of adding an explicit barrier entity to Write batches but decided to postpone it and see if the RTT between batches is actually an issue for anyone, as it can easily be added in the future without breaking backward-compatibility of the current protobuf / gRPC API.

**Comment by jafingerhut** [comment]
Makes sense.  Thinking a bit more, a controller/client may often want to check if there was an error in one batch before issuing the next batch, which requires the RTT.

**Comment by jafingerhut** [comment]
A suggestion for #2, if it doesn't break the complexity bank: An boolean flag in each batch operation indicating whether the client/controller wants the server to force them to be performed in FIFO order in the data plane, vs. it is OK for operations in the batch to take effect in the data plane in an arbitrary order?  Then the client has a way to enforce order even within a batch, if it is important for that batch, or relax that constraint, potentially allowing some servers/agents to go faster on those batches.

**Comment by antoninbas** [comment]
Decision at the 02/21/2018 was to allow the P4Runtime service implementation to re-order operations within a batch when possible. Spec needs to be updated to reflect this.

**Comment by jafingerhut** [comment]
Sorry I missed that part of the meeting.

I would recommend at least considering an option that the controller can choose for a batch operation, between "must do these in the order I gave them" vs. "do in whatever order you think is most efficient".

Why?

Because if the only way to send a batch is "do in whatever order you think is most efficient", then the only recourse a controller has for doing 100 operations in that sequence is to do 100 1-operation batches with an RTT between each.

**Comment by antoninbas** [comment]
See WG decision here: https://docs.google.com/document/d/16gvs3Y196ptz38ujEc9tCGMSm_Elv2I5WdrtEUU3jd0
Closing this issue in favor of #324
